File: src/Kruskal/Kruskal.scala
package kruskal

import graph.Graph
import union_find.UnionFind
import scala.collection.mutable.Set as SET

class Kruskal(graph: Graph){

    var minimumSpanningTree  = SET[(Int, Int, Int)]();

    graph.getGraphType() match {
        case 0 => { //adjacency matrix
            var unionFind = new UnionFind((0 until graph.adjacencyMatrix.length).toArray);

/*             for(i <- unionFind.parent){
                println(i);
            } */
            
            var orderedEdgeList = graph.adjacencyMatrix.zipWithIndex.flatMap{
                                        case (row, x) => (row.zipWithIndex.map{case (value, y) => (x, y, value)})   //first flat map to get the indices and edge weight
                                    }
                                    .sortWith((x, y) => x._3 < y._3)    //sort the edges in increading order
                                    .filter((x) => x._3 != Int.MaxValue);   //remove the infinity (unreachable edge links)

/*             for(i <- orderedEdgeList){
                println(i);
            } */

            createMST(orderedEdgeList, unionFind);
        }


        case 1 => { //adjacency list
            var unionFind = new UnionFind((0 until graph.adjacencyList.length).toArray);

            var orderedEdgeList = graph.adjacencyList.zipWithIndex.flatMap{
                            case (edgeList, x) => (edgeList.map(listItem => (x, listItem._1, listItem._2)))   //first flat map to get the indices and edge weight
                        }
                        .sortWith((x, y) => x._3 < y._3)    //sort the edges in increading order
                        //no need to filter

            createMST(orderedEdgeList, unionFind);
        }

    }

    def createMST(orderedEdgeList: Array[(Int, Int, Int)], unionFindStructure: UnionFind) : Unit = {
        for(i <- orderedEdgeList){
            if(unionFindStructure.find(i._1) != unionFindStructure.find(i._2)){
                minimumSpanningTree = minimumSpanningTree.union(SET((i._1, i._2, i._3)));
                minimumSpanningTree = minimumSpanningTree.union(SET((i._2, i._1, i._3)));

                unionFindStructure.union(unionFindStructure.find(i._1), unionFindStructure.find(i._2));
            }
        }
    }

    def printMinimumSpanningTree() : Unit = {
        for(i <- minimumSpanningTree.toList){
            println(i);
        }
    }
/*     var unionFind = UnionFind(graph);

    //sorting the edges
    var orderedEdgeList = graph.edgeList.sortWith((x, y) => x._3 < y._3);   

    for(i <- 0 until orderedEdgeList.length){
        if(unionFind.find(orderedEdgeList(i)._1) != unionFind.find(orderedEdgeList(i)._2)){

        }
    } */
}

File: Python/Kruskal.py
from Graph import Graph
from UnionFind import UnionFind

class Kruskal:
    def __init__(self, graph):
        self.minimumSpanningTree = set()

        if graph.getGraphType() == 0:  # Adjacency matrix
            unionFind = UnionFind(list(range(len(graph.adjacencyMatrix))))

            orderedEdgeList = [(x, y, value) for x, row in enumerate(graph.adjacencyMatrix) for y, value in enumerate(row)]
            orderedEdgeList.sort(key=lambda x: x[2])
            orderedEdgeList = [edge for edge in orderedEdgeList if edge[2] != float('inf')] # remove edges with weight infinity

            self.createMST(orderedEdgeList, unionFind)

        elif graph.getGraphType() == 1:  # Adjacency list
            unionFind = UnionFind(list(range(len(graph.adjacencyList))))

            orderedEdgeList = [(x, dest, weight) for x, edges in enumerate(graph.adjacencyList) for dest, weight in edges]
            orderedEdgeList.sort(key=lambda x: x[2])

            self.createMST(orderedEdgeList, unionFind)

    def createMST(self, orderedEdgeList, unionFindStructure):
        for edge in orderedEdgeList:
            if unionFindStructure.find(edge[0]) != unionFindStructure.find(edge[1]):
                self.minimumSpanningTree.add((edge[0], edge[1], edge[2]))
                unionFindStructure.union(unionFindStructure.find(edge[0]), unionFindStructure.find(edge[1]))

    def printMinimumSpanningTree(self):
        for edge in self.minimumSpanningTree:
            print(edge)

